### PR TITLE
Refactor of external build to make the output cleaner

### PR DESCRIPTION
**Pull request by cliftonc**
Rolls up output changes from @bettiolo, should make ongoing watching and console output much more usable.

### CODE REVIEW

**Comment by bettiolo** [comment]
If I run `bosco cdn` with this branch, and my build is writing out to stderr, stdout will be displayed but stderr will be omitted. We should output stderr too.

File: src/ExternalBuilders/utils.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      commandForLog = command;
+      var arrayCommand = Array.isArray(command);
+      if (arrayCommand) {
+        commandForLog = JSON.stringify(command);
+        args = command;
+        command = args.shift();
+      }
+    }
+    command = ensureCorrectNodeVersion(command, interpreter);
+    return {command: command, args: args, log: commandForLog, watch: watch, ready: ready, checkDelay: checkDelay, timeout: timeout};
```
Comment (geophree) [inline src/ExternalBuilders/utils.js]: I think we should probably just give up on `arrayCommand` and `args`, we can't really use it with `ensureCorrectNodeVersion()` anyway.  We'll need to migrate the existing ones (if there are any).

File: src/ExternalBuilders/SpawnWatch.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    wc.stderr.on('data', function(data) {
+      if (waitingForOutput) {
+        if (!checkFinishedTimer) checkFinished();
+        waitingForOutput = false;
+        bosco.log('Started build for service ' + service.name.blue + ', but immediately errored ...');
+      }
+      childError = true;
+      errorOutput += data.toString();
+      if (verbose) {
+        process.stdout.write(data.toString());
```
Comment (geophree) [inline src/ExternalBuilders/SpawnWatch.js]: This should be `stderr`.

**Comment by bettiolo** [comment]
I made a substantial refactor, please review in a few minutes

File: src/ExternalBuilders/SpawnWatch.js (hunk 2 above)
Comment (geophree) [inline src/ExternalBuilders/SpawnWatch.js]: Also, I think you need to pull `calledReady` back in, you should stop storing output, and just pipe it to stderr when you've already called the callback.  something like:
```js
if (!calledReady) errorOutput += data.toString();
if (calledReady || verbose) process.stdout.write(data.toString());
```

Comment (geophree) [inline src/ExternalBuilders/SpawnWatch.js]: I meant `stderr`:
```js
if (calledReady || verbose) process.stderr.write(data.toString());
```

File: src/ExternalBuilders/SpawnWatch.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      bosco.error('Watch'.red + ' command for ' + service.name.blue + ' died with code ' + code);
+    });
+
+    wc.stdout.on('data', function(data) {
+      if (waitingForOutput) {
+        if (!checkFinishedTimer) checkFinished();
+        waitingForOutput = false;
+        bosco.log('Started build for service ' + service.name.blue + ' ...');
+      }
+      output += data.toString();
```
Comment (geophree) [inline src/ExternalBuilders/SpawnWatch.js]: Also, I think you need to pull `calledReady` back in, you should stop storing output.  something like:
```js
if (!calledReady) output += data.toString();
```

**Comment by bettiolo** [comment]
@geophree can you review after latest change set?

File: src/ExternalBuilders/SpawnWatch.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+      if (output.length < 1) {
+        buildStarted();
+      }
+
+      output.push({type: type, data: data.toString()});
+      if (verbose) {
+        bosco.process[type].write(data.toString());
+      }
+
+      if (isBuildFinished()) {
```
Comment (bettiolo) [inline src/ExternalBuilders/SpawnWatch.js]: We may want to throttle this by `checkDelay` setting?

File: src/ExternalBuilders/SpawnWatch.js [hunk 5]
```diff
@@ -4,74 +4,74 @@ module.exports = function(bosco) {
...
-      }
-      if (output.indexOf(command.ready) >= 0) {
-        return watchBuildFinished(childError, output);
-      }
-      checkFinishedTimer = setTimeout(checkFinished, command.checkDelay);
-    };
+    function buildCompleted(err) {
+      var rtn = buildFinished(err, output);
+      reset();
+      return rtn;
```
Comment (geophree) [inline src/ExternalBuilders/SpawnWatch.js]: Instead of this way, try this:
```js
var returnedOutput = output;
reset();
buildFinished(err, output);
```

Comment (geophree) [inline src/ExternalBuilders/SpawnWatch.js]: err:
```js
buildFinished(err, returnedOutput);
```

File: src/ExternalBuilders/SpawnWatch.js [hunk 6]
```diff
@@ -4,74 +4,74 @@ module.exports = function(bosco) {
...
-      if (waitingForOutput) {
-        if (!checkFinishedTimer) checkFinished();
-        waitingForOutput = false;
-        bosco.log('Started build for service ' + service.name.blue + ' ...');
+      if (output.length < 1) {
+        buildStarted();
       }
-      output += data.toString();
+
+      output.push({type: type, data: data.toString()});
```
Comment (geophree) [inline src/ExternalBuilders/SpawnWatch.js]: Bring back `calledReady` or equivalent and stop recording output when you've `calledReady`.

File: src/ExternalBuilders/SpawnWatch.js [hunk 7]
```diff
@@ -4,74 +4,74 @@ module.exports = function(bosco) {
...
       if (verbose) {
-        process.stdout.write(data.toString());
+        bosco.process[type].write(data.toString());
```
Comment (geophree) [inline src/ExternalBuilders/SpawnWatch.js]: Also write the output if it's `'stderr'`:
```js
if (verbose || type === 'stderr') {
```

File: src/ExternalBuilders/SpawnWatch.js [hunk 8]
```diff
@@ -4,74 +4,74 @@ module.exports = function(bosco) {
...
       }
-    });
+    }
+
+    function onChildExit(code, signal) {
+      var childError = new Error('Watch process exited with code ' + code + ' and signal ' + signal);
+      childError.code = code;
+      childError.signal = signal;
+      bosco.error('Watch'.red + ' command for ' + service.name.blue + ' died with code ' + code);
+      return buildCompleted(childError);
```
Comment (geophree) [inline src/ExternalBuilders/SpawnWatch.js]: Don't call `buildCompleted()` here if you've already `calledReady`.

File: src/ExternalBuilders/SpawnWatch.js (hunk 6 above)
Comment (bettiolo) [inline src/ExternalBuilders/SpawnWatch.js]: What is the rationale behind this?

File: src/ExternalBuilders/SpawnWatch.js (hunk 7 above)
Comment (bettiolo) [inline src/ExternalBuilders/SpawnWatch.js]: This is there already, does stderr and stdout

File: src/ExternalBuilders/SpawnWatch.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+      }
+
+      output.push({type: type, data: data.toString()});
+      if (verbose) {
+        bosco.process[type].write(data.toString());
+      }
+
+      if (isBuildFinished()) {
+        buildCompleted();
+      }
```
Comment (geophree) [inline src/ExternalBuilders/SpawnWatch.js]: I would move this around a little like so:
```js
var dataString = data.toString();
if (verbose || type === 'stderr') {
  bosco.process[type].write(dataString);
}

if (calledReady) return;

output.push({type: type, data: dataString});
if (isBuildFinished()) buildCompleted();
```

File: src/ExternalBuilders/SpawnWatch.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+      });
+      return outputCache.indexOf(command.ready) >= 0;
+    }
+
+    function onChildOutput(type, data) {
+      if (!data) { return; }
+
+      if (output.length < 1) {
+        buildStarted();
+      }
```
Comment (geophree) [inline src/ExternalBuilders/SpawnWatch.js]: Should also guard this with a `calledReady` check.

File: src/ExternalBuilders/SpawnWatch.js (hunk 9 above)
Comment (bettiolo) [inline src/ExternalBuilders/SpawnWatch.js]: when you are in verbose mode, you output everything. when not in verbose, the stdout and stderr are outputted at the end of the build.

File: src/ExternalBuilders/SpawnWatch.js (hunk 10 above)
Comment (cliftonc) [inline src/ExternalBuilders/SpawnWatch.js]: We removed calledReady as it makes no sense in watch where there are many builds, after each build we now reset, and then do the same thing on subsequent builds vs the old model where the first build was special.

File: src/ExternalBuilders/SpawnWatch.js (hunk 9 above)
Comment (geophree) [inline src/ExternalBuilders/SpawnWatch.js]: I still think we should output `stderr` as it happens anyway.  Regardless, we should stop storing or checking output once we've called the master callback.

**Comment by geophree** [comment]
First pass review done.

File: src/ExternalBuilders/SpawnWatch.js (hunk 10 above)
Comment (geophree) [inline src/ExternalBuilders/SpawnWatch.js]: Ah, so now we're expecting the master callback to be called multiple times?

Comment (cliftonc) [inline src/ExternalBuilders/SpawnWatch.js]: There is a guard on the master next.

Comment (cliftonc) [inline src/ExternalBuilders/SpawnWatch.js]: In the main file after this refactor out to separate files.

**Comment by geophree** [comment]
Ignore all my `calledReady` comments.  I was misunderstanding the point of the refactor.

**Comment by geophree** [comment]
If we're going to be checking for every build within a watch build, we should probably stop watching the files, and pass another callback to reload them every time a build within a watch build finishes.  That would fix the issue where bosco reads the file too soon and it ends up in an inconsistent state.

**Comment by bettiolo** [comment]
@geophree can we keep further improvements outside the scope of this PR and merge it?
/cc @cliftonc

**Comment by cliftonc** [comment]
Published with 0.5.46

**Pull request merged**